lib/spack/spack/environment/environment.py

```diff
@@ -696,8 +696,8 @@ def _rewrite_relative_paths_on_relocation(self, init_file_dir):
             return
 
         for name, entry in self.dev_specs.items():
-            dev_path = entry["path"]
-            expanded_path = os.path.normpath(os.path.join(init_file_dir, entry["path"]))
+            dev_path = substitute_path_variables(entry["path"])
```

Comment:
I think it's fine, although not sure how to deal with `$env`

Reply:
That's a good point. 

Note that `$env` is expanded just when there is an active environment, so it is not enough cloning from its `spack.yaml`.

As you are implicitly saying, among possible expandable variables, `$env` is the only one that still have a "relative" semantic.

Documentation says
> $env: name of the currently active environment

So I can have a `another_env` activated, then I can `spack env create new_one original_env/spack.yaml`, and the `$env` variable in `original_env/spack.yaml` will be expanded to `another_env` and not `original_env`.

Probably using `$env` is the easiest way to loose control of spack environments 😄 but it might be me missing some important usage of it.

Reply:
Actually what should be used is `canonicalize_path` which makes the result absolute. But I just checked and `canonicalize_path` has a bug if the yaml file was opened by relative path itself... then the output is relative. Will look tomorrow.

Reply:
As I said above, I don’t use `canonicalize_path` because it always make paths absolute, and at that point the management of relative paths when copying environments does not have any reason to exist anymore.

Probably you have a broader view since you are working on this relative vs absolute paths in #34958.

Look forward to your comments tomorrow 😉

Reply:
If you read the docstring of `canonicalize_path` that is what we *should* use, but somehow it doesn't live up to its promise in returning an absolute path if you do `spack env create --dir ./new ./old/spack.yaml`... it returns ./old/<dev path>`. That is yet another bug :tada: